### PR TITLE
[Azure Customer Insight] Updated the 2017-01-01 version with the latest interaction API. Added 2017-04-26 version

### PR DESCRIPTION
**Pull request by oucfei**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (veronicagg) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: is "modelAsString: false" intentional? This will make the type an enum, see https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-enum

Comment (veronicagg) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: similar question as previous is "modelAsString: false" intentional?

Comment (veronicagg) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: should this description be different from the one below and more descriptive?

**Review by veronicagg — CHANGES_REQUESTED** [review]
Please review model validation errors at https://travis-ci.org/Azure/azure-rest-api-specs/jobs/246332416#L1358
Operations API is missing from customer-insights spec: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/246332414#L1364
Should 'location' be removed from parameters of PATCH operation https://travis-ci.org/Azure/azure-rest-api-specs/jobs/246332414#L1416 ? Please note that this may cause a breaking change in the SDKs, and therefore a change may not want to be done in the existing api-version but in the new one.
Please review same errors in 2017-04-26 of the spec: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/246332414#L1812
Please review warnings from linter validation as well https://travis-ci.org/Azure/azure-rest-api-specs/jobs/246332414#L1428 for both specs.
For the new api-version, it'd be great if you could have an initial commit of the old version and a newer commit with the changes (see https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md#filenames-and-folder-structure)

**Comment by veronicagg** [comment]
@ravbhatnagar there's a new api-version being added with some small changes, will you be signing off in this review?

**Comment by ravbhatnagar** [comment]
@veronicagg  - yes I would like to. Since its a new file it has to be reviewed using notepad++ or something similar. Any recommendations? Can a PR be sent which compares it against the older api version so that only what changed can be reviewed.

File: arm-customer-insights/2017-01-01/swagger/customer-insights.json
Comment (oucfei) [inline arm-customer-insights/2017-01-01/swagger/customer-insights.json]: Hi, this is intentional and it's OK because the 2017-01-01 version never get released before. Last time when we update this swagger, this part was a work-in-progress. Now that it's done and it should look like this (when you call this API with 2017-01-01 version, this is what the returned schema looks like)

**Comment by oucfei** [comment]
@veronicagg thanks for review the PR. 
The model validation error has been fixed.
Operations API is not implemented yet in customer insight. We have a backlog item for that.
'Location' can not be removed from the PATCH operation because in the code we're explicitly checking it and throws FORBIDDEN if it changed. 
Question: do we have to fix all the warnings from linter validation? Those warnings are already there before this PR.

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: hi, sorry it's unclear which line you're referring to. the "modelAsString: false" appear many times in the swagger. But regardless, those are copied from the old api-version, and no changes in the new api version. So it's already there before this PR. Same for the other places.

Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: sorry which line is this? It doesn't show correctly above

File: arm-customer-insights/2017-01-01/swagger/customer-insights.json
Comment (veronicagg) [inline arm-customer-insights/2017-01-01/swagger/customer-insights.json]: If the current spec does not match the wire, meaning that it was initially incorrect, then it may be ok corrected (though for anyone who shipped it avoiding a breaking change is best). Regarding your statement about "201-01-01 version never get released before", what do you mean by this? 
Having the swagger spec in master of this repo, means anyone could generate an SDK, I see there's an SDK for Go at https://github.com/Azure/azure-sdk-for-go/tree/master/arm/customer-insights.

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (veronicagg) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: I'm referring to targetEntityType, sourceEntityType which are new properties in this spec.

Comment (veronicagg) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: descriptions of sourceEntityType and targetEntityType are the same.

**Comment by veronicagg** [comment]
@ravbhatnagar I've been using VS Code, and comparing with the old version, see http://dailydotnettips.com/2015/06/04/how-to-compare-files-in-visual-studio-code/
We do want the old version of the spec in the first commit and then commits on top of that to make it easier for review, and I asked the author about it, see above. In the meantime, you can use VS Code.

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: This is intentional. They're indeed enum type

Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: fixed it

Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: same as above

File: arm-customer-insights/2017-01-01/swagger/customer-insights.json
Comment (oucfei) [inline arm-customer-insights/2017-01-01/swagger/customer-insights.json]: hi sorry looks like it's released before. However this part was incomplete so I'm sure no one was using it. Anyway the version 2017-01-01 should look like this swagger now

**Comment by veronicagg** [comment]
@oucfei I've replied to your questions in the inline comments, please review. 
Regarding your points above:
1. Thanks, please make sure the examples are matching the wire format
2. Ok, be aware that until fixed, it'll get reported as a violation. @ravbhatnagar - FYI this is an RPC violation.
3. 'Location' is required in the body parameter of the PATCH operation at the moment, so are you requiring  users of this API to send the property with the same value as it was originally set? @ravbhatnagar RPC violdation https://travis-ci.org/Azure/azure-rest-api-specs/jobs/246332414#L1416
4. We prefer it this way yes. You could change your previous commits, yes, though if there are review comments they will get messed up, so you can do it after the current comments, but I wouldn't do it after @ravbhatnagar reviews the spec, or they'll get kind of lost. 

RE: Questions:
1. Linter validation: Since you're introducing a new api-version, it's a good time to review all violations. Note that "errors" will be reported for the spec, "warnings" are strong recommendations, you should evaluate the ones that make sense to you.

2. Travis CI - there's an issue with one of the tools, I've reported it and the owner is working on it, it shouldn't be due to this PR, we expect it'll be resolved soon.

**Comment by oucfei** [comment]
@veronicagg thanks for your reply
1. I have ran "oav validation-example" and no error found.
2. OK.
3. I looked at the code again, turned out the location is not required. I have removed from the required properties in the swagger.
4. OK I will not modify my previous commits for now.

I also looked at the warnings from linter validation and they don't make sense to me, so I will not fix those.

**Comment by veronicagg** [comment]
@oucfei thanks. Regarding #1, ok, the tool will compare the examples with the spec, so if the wire format matches the example, then the spec is more accurately representing the wire format. 
#3. Ok, I'll let @ravbhatnagar comment on it, since a "Resource" becomes a "Proxy resource" if location is not required, and I'd like to make sure from ARM perspective that's appropriate for your spec. 

@ravbhatnagar please review and provide your comments/approval, on behalf of ARM team.

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (ravbhatnagar) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: Existing API but still wanted to note that resource types should be pluralized - kpi

Comment (ravbhatnagar) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: This resource type name conflicts with the Microsoft.Authorization/roleAssignments which is a commonly known RBAC model. This should not have been named this since we dont want to confuse users with the Azure wide concepts like RBAC role assignments.

Comment (ravbhatnagar) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: Segments not correct for this API. It would be working for you since ARM may be routing based on the parent. The correct segments should be /images/{imageName}/getEntityTypeImageUploadUrl. If imageName is not applicable then do you really need the "images" type? Can you invoke this action on a hubs/{hubName}

**Review by ravbhatnagar — CHANGES_REQUESTED** [review]
Please take a look at the comments I have added. Majority of those are around existing APIs. 
I see in a lot of resource types you have "name" top level property and also a name property inside the properties bag. It is redundant information.

**Comment by ravbhatnagar** [comment]
For #3 - Location should be required for Tracked resources. All this means is that your create parameters look different from your update parameters. You just have to add another model which becomes the request body of your PATCH. That way customers know what is required and what is not and also what can or cannot be updated.

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: hi, I don't know any specific reason except that this is how the API was designed and implemented. At this point we're not planning to make change to this API.

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: same as above.

**Comment by veronicagg** [comment]
@oucfei Please review @ravbhatnagar comments, and if you confirm the resources mentioned in "3" are tracked resources, please make the corresponding updates to the spec. Thanks!

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: I agree it should be sufficed with just one API, but again this is how the API was designed and implemented, and I don't know why.

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: not sure which line you're referring to

Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: in our API the "tenantId" is actually the "hubId"... we have a bad name and it is confusing...

**Comment by oucfei** [comment]
@ravbhatnagar thanks for your review. 
1. Most of the comment you left are on existing APIs, and I don't know how to resolve those except to tell you that this is how the API was designed and implemented. We're not going to make change to the API, nor the swagger at this point of time. Anyway this swagger was reviewed and approved before, so I think those issues should not block the PR, right?
2. Is it just me or the same for others, that I can't see which lines of code you're referring to in some of your comments? Again this has already been approved before so I don't think there're any real issues, but if you don't agree, please leave the exact line numbers you're referring to, and I'll take a look again.
3. I'm not 100% sure what a Tracked resources means, but the create parameters and the update parameters for this API are the same. There's no additional model for PATCH. So I guess it's not tracked resource?

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: which line is this?

Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: I don't see it already present in the "name" property... which line is that?

Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: which line is this?

Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: which line is this?

Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: I don't know why it doesn't have {imageName}, but it is what our API looks like and it seems working fine

Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: which line is this? But anyway I don't think we can change this now.

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (ravbhatnagar) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: Ok. consider changing this in the next API-version revision to make the API cleaner.

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (ravbhatnagar) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: Not sure why it doesnt show up there. But likely you wont be able to change it now. I am referring to the resource type "kpi"

Comment (ravbhatnagar) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: Not sure why the line numbers got messed up for some comments. But likely you wont be able to change it now. I am referring to the resource type "roleAssignments"

Comment (ravbhatnagar) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: What does this API do? Is the URL same for all images in a hubs/{hubName}. I am wondering if its working as you expect then why do you even need a /images segment?

**Comment by ravbhatnagar** [comment]
Yes, you are right. Feel free to ignore the feedback around existing APIs. We will follow up with you separately on that. Tracked resource means that the "routingType" is set to "default" in the ARM manifest. Hubs is a tracked resource. Anyway location of a resource cant be changed. And so if someone tries to PATCH (change the location) which they can since the same model is being used, the operation will/should fail.

File: arm-customer-insights/2017-04-26/swagger/customer-insights.json
Comment (oucfei) [inline arm-customer-insights/2017-04-26/swagger/customer-insights.json]: Hi, this api is to get the images' upload URL. As you can see in the swagger it depends on the entityType, entityTypeName, relativePath, so the URL is not the same for all images in a hub, but it has nothing to do with the imageName. This API is in the ImagesController, I think that's why the person who developed this add a /images segment

**Comment by oucfei** [comment]
@veronicagg @ravbhatnagar I've resolved all the comments, and we've agreed not to fix the issues around existing API at this time. So could you please sign off this PR? Please let me know if you still have questions.

**Comment by ravbhatnagar** [comment]
@oucfei - Only one open question - Tracked resource means that the "routingType" is set to "default" in the ARM manifest. Hubs is a tracked resource. Anyway location of a resource cant be changed. And so if someone tries to PATCH (change the location) which they can since the same model is being used, the operation will/should fail. @veronicagg - This probably is not ideal SDK experience. Please confirm.

**Comment by oucfei** [comment]
@ravbhatnagar got it. Made "location" required in the PATCH operation.

**Comment by ravbhatnagar** [comment]
@oucfei - Location of a resource cannot be updated after the resource has been created. So it cant be made required in the PATCH. Ideally, it should not be present in the request body of the PATCH.

**Comment by oucfei** [comment]
@ravbhatnagar  OK... sorry for my misunderstanding. But currently in the API, location is part of the patch request body, and we're explicitly checking if it's changed. So I'm not going to make change to the swagger at this time...

**Comment by oucfei** [comment]
@ravbhatnagar do you agree with my above comment? can you sign off now?

**Comment by ravbhatnagar** [comment]
Ok, but please consider updating this at some point since it will improve the client experience. 
@veronicagg - ARM signs off.

**Comment by veronicagg** [comment]
ok, thanks @ravbhatnagar . 
@oucfei just so you know, from the SDK perspective, having location not required and as shared model for put and patch will result in an SDK API that won't "require" location for PUT body parameter (https://github.com/oucfei/azure-rest-api-specs/blob/9a36dc27b6b6a4470fad4c39d55f18161178f017/arm-customer-insights/2017-01-01/swagger/customer-insights.json#L75) so the consumer of the API may not send location and get a failed response from the service. For next updates to the API, you'd want to consider creating separate models for PUT or PATCH so you can more accurately represent the service.

**Review by veronicagg — APPROVED** [review]

**Comment by oucfei** [comment]
@veronicagg thanks! I will create items for the issues mentioned above.

**Pull request merged**
